### PR TITLE
experimental services ui

### PR DESCRIPTION
**Pull request by mavenugo**
Based on various discussions and inputs, we have a simplified proposal to address the network UI
separate from services ui. This PR brings in the services ui & api provided by libnetwork and hooking it
to be part of docker.
All of these are defined under experimental tag in order to get user feedback before making it officially into the docker ui chain.

### CODE REVIEW

**Comment by jfrazelle** [comment]
LGTM

File: daemon/container_linux.go [hunk 1]
```diff
@@ -763,31 +775,51 @@ func (container *Container) AllocateNetwork() error {
 		return nil
 	}
 
+	var networkDriver string
+	service := container.Config.PublishService
 	networkName := mode.NetworkName()
+	canCreateNetwork := false
```
Comment (LK4D4) [inline daemon/container_linux.go]: What the point of this variable? This is basically `mode.IsDefault()`

File: daemon/container_linux.go [hunk 2]
```diff
@@ -763,31 +775,51 @@ func (container *Container) AllocateNetwork() error {
...
 			return err
 		}
 	}
 
-	createOptions, err := container.buildCreateEndpointOptions()
+	ep, err := n.EndpointByName(service)
 	if err != nil {
-		return err
-	}
+		createOptions, err := container.buildCreateEndpointOptions()
```
Comment (LK4D4) [inline daemon/container_linux.go]: Looks pretty weird that we do this on ANY error. At least need comments.

File: daemon/container_linux.go [hunk 3]
```diff
@@ -763,31 +775,51 @@ func (container *Container) AllocateNetwork() error {
...
 	if mode.IsDefault() {
-		networkName = controller.Config().Daemon.DefaultNetwork
+		canCreateNetwork = true
+		if service != "" {
+			service, networkName, networkDriver = parseService(controller, service)
+		} else {
+			networkName = controller.Config().Daemon.DefaultNetwork
+			networkDriver = controller.Config().Daemon.DefaultDriver
+		}
+	} else {
```
Comment (LK4D4) [inline daemon/container_linux.go]: `else if service != ""`

**Comment by LK4D4** [comment]
I think most interesting changes here in docs.
ping @moxiegirl @jamtur01 @SvenDowideit

File: experimental/networking.md [hunk 4]
```diff
@@ -59,14 +62,53 @@ If you no longer have need of a network, you can delete it with `docker network
...
-Currently the only way this network can be used to connect container is via default network-mode.
 Docker daemon supports a configuration flag `--default-network` which takes configuration value of format `NETWORK:DRIVER`, where,
 `NETWORK` is the name of the network created using the `docker network create` command and 
 `DRIVER` represents the in-built drivers such as bridge, overlay, container, host and none. or Remote drivers via Network Plugins.
 When a container is created and if the network mode (`--net`) is not specified, then this default network will be used to connect
 the container. If `--default-network` is not specified, the default network will be the `bridge` driver.
 
+## Using Services
+
+        Usage: docker service [OPTIONS] COMMAND [OPTIONS] [arg...]
```
Comment (LK4D4) [inline experimental/networking.md]: what's first `[OPTIONS]`?

File: experimental/networking.md [hunk 5]
```diff
@@ -2,7 +2,10 @@
 
 In this feature:
 
-- `network` become a first class objects in the Docker UI
+- `network` and `service` become a first class objects in the Docker UI
+- You can create networks and attach containers to them
+- We introduce the concept of `services`
```
Comment (moxiegirl) [inline experimental/networking.md]: period at the end of the sentence.

**Comment by moxiegirl** [comment]
LGTM (networking.md to be sorted globally at a later date.)

File: daemon/container_linux.go (hunk 1 above)
Comment (mavenugo) [inline daemon/container_linux.go]: Not required. it was added for readability.
But have changed it (added comments instead).

File: daemon/container_linux.go (hunk 2 above)
Comment (mavenugo) [inline daemon/container_linux.go]: Good catch. It must be checking for `libnetwork.ErrNoSuchEndpoint`. Fixed it.

File: daemon/container_linux.go (hunk 3 above)
Comment (mavenugo) [inline daemon/container_linux.go]: done.

**Comment by icecrime** [comment]
LGTM

**Pull request merged**